exchanges/fee/commission.go

```diff
@@ -0,0 +1,234 @@
+	}
+}
+
+// CalculateMaker returns the calculated maker fees
+func (c *CommissionInternal) CalculateMaker(price, amount float64) (float64, error) {
+	if price == 0 {
+		return 0, errPriceIsZero
+	}
+	if amount == 0 {
+		return 0, errAmountIsZero
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.calculate(c.maker, price, amount)
+}
+
+// CalculateTaker returns the calculated taker fees
+func (c *CommissionInternal) CalculateTaker(price, amount float64) (float64, error) {
+	if price == 0 {
+		return 0, errPriceIsZero
+	}
+	if amount == 0 {
+		return 0, errAmountIsZero
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.calculate(c.taker, price, amount)
+}
+
+// CalculateWorstCaseMaker returns the worst-case calculated maker fees
+func (c *CommissionInternal) CalculateWorstCaseMaker(price, amount float64) (float64, error) {
+	if price == 0 {
+		return 0, errPriceIsZero
+	}
+	if amount == 0 {
+		return 0, errAmountIsZero
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.calculate(c.worstCaseMaker, price, amount)
+}
+
+// CalculateWorstCaseTaker returns the worst-case calculated taker fees
+func (c *CommissionInternal) CalculateWorstCaseTaker(price, amount float64) (float64, error) {
+	if price == 0 {
+		return 0, errPriceIsZero
+	}
+	if amount == 0 {
+		return 0, errAmountIsZero
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.calculate(c.worstCaseTaker, price, amount)
+}
+
+// GetMaker returns the maker fee and type
+func (c *CommissionInternal) GetMaker() (fee float64, isFixedAmount bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.maker.InexactFloat64(), c.isFixedAmount
+}
+
+// GetTaker returns the taker fee and type
+func (c *CommissionInternal) GetTaker() (fee float64, isFixedAmount bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.taker.InexactFloat64(), c.isFixedAmount
+}
+
+// GetWorstCaseMaker returns the worst-case maker fee and type
+func (c *CommissionInternal) GetWorstCaseMaker() (fee float64, isFixedAmount bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.worstCaseMaker.InexactFloat64(), c.isFixedAmount
+}
+
+// GetWorstCaseTaker returns the worst-case taker fee and type
+func (c *CommissionInternal) GetWorstCaseTaker() (fee float64, isFixedAmount bool) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.worstCaseTaker.InexactFloat64(), c.isFixedAmount
+}
+
+// set sets the commission values for update
+func (c *CommissionInternal) set(maker, taker float64, isFixedAmount bool) error {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	if c.isFixedAmount != isFixedAmount {
+		// These should not change, and a package update might need to occur.
+		return errFeeTypeMismatch
+	}
+	c.maker = decimal.NewFromFloat(maker)
+	c.taker = decimal.NewFromFloat(taker)
+	return nil
+}
+
+// calculate returns the commission fee total based on internal loaded values
+func (c *CommissionInternal) calculate(fee decimal.Decimal, price, amount float64) (float64, error) {
+	// TODO: Add fees based on volume of this asset
```

Comment:
I think this function can separate from `(c *CommissionInternal)`, you keep readlocking before calling this function, but you pass in everything save `c.isFixedAmount`, which you can pass in at the same time you're passing in fee. Such as:
```
func calculate(isFixedRate bool, fee decimal.Decimal, price, amount float64) (float64, error) {
```

The reason I say that is because there's a risk here by not calling lock here that its not guaranteed to be locked. It can still keep ``(c *CommissionInternal)`, but yeah passing in the bool will be beneficial here